src/SuperVault.sol

```diff
@@ -430,54 +441,127 @@ contract SuperVault is BaseStrategy, ISuperVault {
     //////////////////////////////////////////////////////////////
     //                     INTERNAL FUNCTIONS                   //
     //////////////////////////////////////////////////////////////
-
+    struct PrepareMultiVaultDataLocalVars {
+        uint256 totalAssetsInVaults;
+        uint256[] spBalances;
+        uint256[] assetBalances;
+    }
     /// @notice Prepares multi-vault data for deposit or withdrawal
     /// @param amount_ The amount to deposit or withdraw
-    /// @param isDeposit True if depositing, false if withdrawing
+    /// @param isDeposit_ True if depositing, false if withdrawing
     /// @return mvData The prepared multi-vault data
+
     function _prepareMultiVaultData(
         uint256 amount_,
-        bool isDeposit
+        bool isDeposit_
     )
         internal
         view
         returns (MultiVaultSFData memory mvData)
     {
-        uint256 numberOfSuperforms = numberOfSuperforms;
-
-        mvData.superformIds = superformIds;
-        mvData.amounts = new uint256[](numberOfSuperforms);
-        mvData.maxSlippages = new uint256[](numberOfSuperforms);
-        mvData.liqRequests = new LiqRequest[](numberOfSuperforms);
-        mvData.hasDstSwaps = new bool[](numberOfSuperforms);
+        uint256 _numberOfSuperforms_ = numberOfSuperforms;
+        uint256[] memory _superformIds_ = superformIds;
+
+        mvData.superformIds = _superformIds_;
+        mvData.amounts = new uint256[](_numberOfSuperforms_);
+        mvData.maxSlippages = new uint256[](_numberOfSuperforms_);
+        mvData.liqRequests = new LiqRequest[](_numberOfSuperforms_);
+        mvData.hasDstSwaps = new bool[](_numberOfSuperforms_);
         mvData.retain4626s = mvData.hasDstSwaps;
         mvData.receiverAddress = address(this);
         mvData.receiverAddressSP = address(this);
-        mvData.outputAmounts = new uint256[](numberOfSuperforms);
+        mvData.outputAmounts = new uint256[](_numberOfSuperforms_);
 
         /// @dev caching to avoid multiple MLOADs
         address superform;
         IBaseForm superformContract;
 
-        for (uint256 i; i < numberOfSuperforms; ++i) {
+        PrepareMultiVaultDataLocalVars memory vars;
+        vars.totalAssetsInVaults = 0;
+        vars.spBalances = new uint256[](_numberOfSuperforms_);
+        vars.assetBalances = new uint256[](_numberOfSuperforms_);
+
+        // 1. Snapshot assets and SP balances. This is relevant for withdraws and must be calculated in advance
+        for (uint256 i; i < _numberOfSuperforms_; ++i) {
+            (superform,,) = _superformIds_[i].getSuperform();
+            superformContract = IBaseForm(superform);
+
+            vars.spBalances[i] =
+                ISuperPositions(_getAddress(keccak256("SUPER_POSITIONS"))).balanceOf(address(this), _superformIds_[i]);
```

Comment:
`_getAddress(keccak256("SUPER_POSITIONS")))` can be cached to improve readbility and safe gas

Reply:
[40ee60d](https://github.com/superform-xyz/SuperVaults/pull/36/commits/40ee60d2be8c582e796a278e1f1b7d64a3221194)